hal/armv8m/exceptions.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Exception handling
+ *
+ * Copyright 2017, 2022 Phoenix Systems
+ * Author: Pawel Pisarczyk, Jakub Sejdak, Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "../exceptions.h"
+#include "../cpu.h"
+#include "../console.h"
+#include "../string.h"
+#include "config.h"
+
+
+void hal_exceptionsDumpContext(char *buff, exc_context_t *ctx, int n)
+{
+	static const char *mnemonics[] = {
+		"0 #InitialSP", "1 #Reset", "2 #NMI", "3 #HardFault",
+		"4 #MemMgtFault", "5 #BusFault", "6 #UsageFault", "7 #SecureFault",
+		"8 #", "9 #", "10 #", "11 #SVC",
+		"12 #Debug", "13 #", "14 #PendSV", "15 #SysTick"
+	};
+	size_t i = 0;
+	u32 msp = (u32)ctx + sizeof(*ctx);
+	u32 psp_s = ctx->psp_s;
+	cpu_hwContext_t *hwctx;
+
+	/* If we came from userspace HW ctx in on psp stack */
+	if (ctx->excret == RET_THREAD_PSP) {
+		hwctx = (void *)ctx->psp_s;
+		msp -= sizeof(cpu_hwContext_t);
+		psp_s += sizeof(cpu_hwContext_t);
+	}
+	else {
+		hwctx = &ctx->mspctx;
+	}
+
+	n &= 0xf;
+
+	hal_strcpy(buff, "\nException: ");
+	i = sizeof("\nException: ") - 1;
+	hal_strcpy(&buff[i], mnemonics[n]);
+	i += hal_strlen(mnemonics[n]);
+
+	i += hal_i2s("\n r0=", &buff[i], hwctx->r0, 16, 1);
+	i += hal_i2s("  r1=", &buff[i], hwctx->r1, 16, 1);
+	i += hal_i2s("  r2=", &buff[i], hwctx->r2, 16, 1);
+	i += hal_i2s("  r3=", &buff[i], hwctx->r3, 16, 1);
+
+	i += hal_i2s("\n r4=", &buff[i], ctx->r4, 16, 1);
+	i += hal_i2s("  r5=", &buff[i], ctx->r5, 16, 1);
+	i += hal_i2s("  r6=", &buff[i], ctx->r6, 16, 1);
+	i += hal_i2s("  r7=", &buff[i], ctx->r7, 16, 1);
+
+	i += hal_i2s("\n r8=", &buff[i], ctx->r8, 16, 1);
+	i += hal_i2s("  r9=", &buff[i], ctx->r9, 16, 1);
+	i += hal_i2s(" r10=", &buff[i], ctx->r10, 16, 1);
+	i += hal_i2s(" r11=", &buff[i], ctx->r11, 16, 1);
+
+	i += hal_i2s("\nr12=", &buff[i], hwctx->r12, 16, 1);
+	i += hal_i2s(" psr=", &buff[i], hwctx->psr, 16, 1);
+	i += hal_i2s("  lr=", &buff[i], hwctx->lr, 16, 1);
+	i += hal_i2s("  pc=", &buff[i], hwctx->pc, 16, 1);
+
+	i += hal_i2s("\npsp_s=", &buff[i], psp_s, 16, 1);
```

Comment:
Should it  be named with _s suffix? At least for now we have only non-secure mode, so according to DDI0553B.r B3.8 it should be named without _s/_ns suffix. And you don't add this suffix to msp.

Reply:
Yeah, maybe it would be better, I wanted to show that there is only secure mode in our implementation, but let's be consistent with the docs.

Reply:
Welp, It bring my attention mainly because it causes that exception will be misformated.